generators/typescript/codegen/src/ast/core/Writer.ts

```diff
@@ -1,7 +1,78 @@
 import { AbstractWriter } from "@fern-api/generator-commons";
+import { Reference } from "../Reference";
+
+type ModuleName = string;
+
+export declare namespace Writer {
+    interface Args {
+        /* The name of the module that is being written to */
+        moduleName: string;
+    }
+}
 
 export class Writer extends AbstractWriter {
+    protected imports: Record<ModuleName, Reference[]> = {};
+    protected defaultImports: Record<ModuleName, Reference> = {};
+
+    /**
+     * Adds the given import under its module name.
+     */
+    public addImport(reference: Reference): void {
+        if (reference.module != null) {
+            if (reference.module.defaultExport ?? false) {
+                const existing = this.defaultImports[reference.module.moduleName];
+                if (existing == null) {
+                    this.defaultImports[reference.module.moduleName] = reference;
+                } else if (existing.name !== reference.name) {
+                    throw new Error(
+                        `Cannot have multiple default imports for module ${reference.module.moduleName}: ` +
+                            `got ${reference.name} but already had ${existing.name}`
+                    );
+                }
+            }
+            this.imports[reference.module.moduleName] ??= [];
+            const moduleImports = this.imports[reference.module.moduleName];
+            if (moduleImports != null) {
+                const names = moduleImports.map((import_) => import_.name);
+                if (!names.includes(reference.name)) {
+                    moduleImports.push(reference);
+                }
+            }
+        }
+    }
```

Comment:
Thoughts on splitting this up a bit to reduce nesting? It's otherwise a little hard to follow at first glance given that there's a variety of conditions. What about something along the lines of the following? 

```typescript
public addImport(reference: Reference): void {
  if (reference.module == null) {
    return;
  }
  if (reference.module.defaultExport) {
    this.addDefaultExportOrThrow({ reference, moduleName: reference.module.name } );
    return;
  }
  const references = this.imports[reference.module.moduleName] ??= [];
  if (!references.includes(reference.name)) {
    references.push(reference);
  }
}
```

```typescript
private addReferenceOrThrow({ reference, moduleName }: { reference: Reference; moduleName: string; }): void {
  if (this.defaultImports[moduleName] != null) {
    throw new Error("Internal error; cannot have multiple default imports for module ...");
  }
  this.defaultImports[moduleName] = reference;
}
```

Reply:
done

Reply:
Split it up into the different conditions

Reply:
Ended up looking quite different than to this because I added stars